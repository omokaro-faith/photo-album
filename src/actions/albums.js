import axios from 'axios';
import swal from 'sweetalert';
import { ERROR_OCCURED } from '../constants/constants';

export const setAlbums = albums => ({
  type: 'GET_ALBUMS',
  albums,
});

export const setAlbumsError = () => ({
  type: 'GET_ALBUMS_ERROR',
  message: 'Cannot fetch albums'
});

export const setAlbumsLength = totalAlbums => ({
  type: 'GET_ALL_ALBUMS',
  totalAlbums,
});

export const fetchAllAlbumsError = () => ({
  type: 'GET_ALL_ALBUMS_ERROR',
  errorMessage: 'Cannot fetch all albums'
});

export const fetchAllAlbums = () => async (dispatch) => {
  try {
    const albums = await axios.get('https://jsonplaceholder.typicode.com/albums');
    return dispatch(setAlbumsLength(albums.data.length));
  } catch (err) {
    const { error, status, buttons } = ERROR_OCCURED;
    swal(error, 'Cannot fetch all albums', status, {
				buttons,
		});
    return(dispatch(fetchAllAlbumsError()));
  }
}

export const getAlbums = (start = 0, limit) => async (dispatch) => {
  try {
    const albums = await axios.get(`https://jsonplaceholder.typicode.com/albums?_start=${start}&_limit=${limit}`);
    return dispatch(setAlbums(albums.data));
  } catch (err) {
    const { error, status, buttons } = ERROR_OCCURED;
    swal(error, 'Cannot fetch albums', status, {
				buttons,
		});
    return(dispatch(setAlbumsError()));
  }
};
