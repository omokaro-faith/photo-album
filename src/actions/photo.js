import axios from 'axios';

export const setPhotos = photos => ({
  type: 'GET_PHOTOS',
  photos,
});

export const setPhotoError = () => ({
  type: 'GET_PHOTOS_ERROR',
  message: true
});

export const setPhotosLength = totalPhotos => ({
  type: 'GET_ALL_PHOTOS',
  totalPhotos,
});

export const fetchAllPhotosError = () => ({
  type: 'GET_ALL_PHOTOS_ERROR',
  errorMessage: true
});

export const fetchAllPhotos = albumId => async (dispatch) => {
  try {
    const photos = await axios.get(`https://jsonplaceholder.typicode.com/photos?albumId=${albumId}`);
    return dispatch(setPhotosLength(photos.data.length));
  } catch (err) {
    return(dispatch(fetchAllPhotosError()));
  }
}

export const getPhotos = (albumId, start= 0 , limit) => async (dispatch) => {
  try {
    const photos = await axios.get(`https://jsonplaceholder.typicode.com/photos?albumId=${albumId}&_start=${start}&_limit=${limit}`);
    return dispatch(setPhotos(photos.data));
  } catch (err) {
    return(dispatch(setPhotoError()));
  }
};
