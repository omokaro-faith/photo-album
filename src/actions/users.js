import axios from 'axios'

export const setUsers = (users) => ({
 type: 'GET_USERS',
 users
});

export const getUsers = () => {
 return async dispatch => {
   try {
    const users = await axios.get('https://jsonplaceholder.typicode.com/users')
    return dispatch(setUsers(users.data));
   } catch (error) {
     console.log(error);
   }
 }
};
