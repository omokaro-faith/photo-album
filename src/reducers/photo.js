const initialState = {isLoading: true}

 export default (state= initialState, action) => {
 switch (action.type) {
   case 'GET_PHOTO':
   const { users } = action
     return {...state, users, isLoading: false};
   default:
     return initialState;
 }
}