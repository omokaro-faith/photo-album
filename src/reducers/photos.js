const initialState = {}

 export default (state= initialState, action) => {
 switch (action.type) {
   case 'GET_PHOTOS':
   const { photos } = action;
     return {...state, photos};
    case 'GET_ALL_PHOTOS':
      const { totalPhotos } = action;
    return {...state, totalPhotos};
   case 'GET_PHOTOS_ERROR':
   const { message } = action; 
     return { ...state, message };
   case 'GET_ALL_PHOTOS_ERROR':
      const { errorMessage } = action; 
      return { ...state, errorMessage}
   default:
     return state;
 }
}
