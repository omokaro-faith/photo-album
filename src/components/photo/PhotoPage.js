import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Dropdown from '../pagination/Dropdown';
import Buttons from '../pagination/Button';
import Modal from '../modal/Modal';
import { getPhotos }  from '../../actions/photo';
import Header from '../header/Header';
import { getPageNumbers, getCurrentItems } from '../../utils/utils';

export class PhotoPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false,
      modalContent: {},
      currentPage: 1,
      itemsPerPage: 10,
    };
    this.handleChange = this.handleChange.bind(this);
    this.showModal = this.showModal.bind(this);
    this.handleClick = this.handleClick.bind(this);
  };

  static propTypes = {
    getPhotos: PropTypes.func.isRequired,
    photos: PropTypes.array.isRequired,
    match: PropTypes.shape({
      params: PropTypes.shape({
        albumOwner: PropTypes.string.isRequired,
        albumTitle: PropTypes.string.isRequired
      })
    }).isRequired
  };

  componentDidMount() {
    const { match: { params }, getPhotos } = this.props;
    getPhotos(undefined, params.albumId)
  }

  handleChange(event) {
    const { match: { params }, getPhotos } = this.props;
    getPhotos(event.target.value, params.albumId);
    this.setState({
      currentPage: 1,
    });
  }

  handleClick(event) {
    this.setState({
      currentPage: Number(event.target.id)
    });
  }

  showModal = ({ content }) => {
    this.setState({ 
      show: true,
      modalContent: {
        picture: content.url,
        title: content.title,
      }
     });
  };

  hideModal = () => {
    this.setState({ 
      show: false,
      modalContent: {},
     });
  };

  render() {
    const { photos, match: { params: { albumOwner, albumTitle } } } = this.props;
    const { show, modalContent, currentPage, itemsPerPage } = this.state;
    const renderPhotos = getCurrentItems(currentPage, itemsPerPage, photos).map((photo, index) => 
    (
      <div key={index} className="grid__item" onClick={() => this.showModal({ content: photo })}>
        <div className="grid__img">
            <img src={`${photo.thumbnailUrl}`} alt='' title='' className='grid__img--item'/>
        </div>
        <h5>Title: {photo.title}</h5>
      </div>
    ));

    const renderPageNumbers = getPageNumbers(photos, itemsPerPage).map((number) => (
      <Buttons 
       key={number}
       id={number}
       handleClick={this.handleClick}
       pageNumber={number}
      />
   ))


    return (
      <section className="grid">
        <Header totalItems={photos.length} pageName='Photos'/>
        <div className='dropdown-button__wrapper'>
         <Dropdown handleChange={this.handleChange} itemsLength={photos.length}/>
         {renderPageNumbers}
        </div>
        { photos.length >= 1 ? 
          <Fragment>
            <h3>Owner: {albumOwner}</h3>
            <h3>Album Title: {albumTitle}</h3>
            <div className="grid__container">
            { renderPhotos }
            </div>
          </Fragment>
        : <div className="loader"></div> }
        <Modal show={show} handleClose={this.hideModal}>
            <div className='img-content'>
              <img src={`${modalContent.picture}`} alt='' title=''/>
            </div>
            <div>
              <h5>Owner: <span>{albumOwner}</span></h5>
              <h5>Title: {modalContent.title}</h5>
            </div>
        </Modal>
      </section> 
    )
  }
}

PhotoPage.defaultProps = {
  photos: [],
};

const mapDispatchToProps = (dispatch) => ({
  getPhotos: (limit, albumId) => dispatch(getPhotos(limit, albumId)),
});

 const mapStateToProps = (state) => ({
  photos: state.photos.photos,
});



export default connect(mapStateToProps, mapDispatchToProps)(PhotoPage);
