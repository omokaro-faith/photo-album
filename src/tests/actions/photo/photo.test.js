import configureMockStore from 'redux-mock-store'
import nock from 'nock';
import thunk from 'redux-thunk';
import { setPhotos, getPhotos, setPhotoError, setPhotosLength, fetchAllPhotosError, fetchAllPhotos } from '../../../actions/photo';
import photos from '../../fixtures/photos';

const createMockStore = configureMockStore([thunk]);

describe('setPhotos', () => {
  it('should set photos', () => {
    const action = setPhotos(photos);

  expect(action).toEqual({
    type: 'GET_PHOTOS',
    photos,
  });
  });
});

describe('setPhotosError', () => {
  it('should set photos error if an error occurs while setting photos', () => {
    const action = setPhotoError(photos);
    const expected = {
      type: 'GET_PHOTOS_ERROR',
      message: 'Cannot fetch photos'
    }

  expect(action).toEqual(expected);
  });
});

describe('setPhotosLength', () => {
  it('should set photos length', () => {
    const totalPhotos = photos.length;
    const action = setPhotosLength(totalPhotos);
    const expected = {
      type: 'GET_ALL_PHOTOS',
      totalPhotos
    }

  expect(action).toEqual(expected);
  });
});

describe('fetchAllPhotosError', () => {
  it('should set photos error if an error occurs while fetching all photos', () => {
    const action = fetchAllPhotosError(photos);
    const expected = {
      type: 'GET_ALL_PHOTOS_ERROR',
      errorMessage: 'Cannot fetch all photos'
    }

  expect(action).toEqual(expected);
  });
});


describe('Async getPhotos', () => {
  let store;

  beforeEach(() => {
    store = createMockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('gets photos',async () => {
    nock('https://localhost:8000')
      .get('jsonplaceholder.typicode.com/photos?albumId=4&_start=0&_limit=')
      .reply(200);


    await store.dispatch(getPhotos())
      .then(() => {
        expect(store.getActions()[0].type).toEqual('GET_PHOTOS');
        expect(store.getActions()).toMatchSnapshot();
      })
      .catch(() => {
        expect(store.getActions()[0].type).toEqual('GET_PHOTOS_ERROR');
        expect(store.getActions()[0].message).toEqual('Cannot fetch photos');
      });
  });
})

describe('Async fetchAllPhotos', () => {
  let store;

  beforeEach(() => {
    store = createMockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('fetch all photos',async () => {
    nock('https://localhost:8000')
      .get('jsonplaceholder.typicode.com/photos?albumId=2')
      .reply(200);


    await store.dispatch(fetchAllPhotos(2))
      .then(() => {
        expect(store.getActions()[0].type).toEqual('GET_ALL_PHOTOS');
        expect(store.getActions()).toMatchSnapshot();
      })
      .catch(() => {
        expect(store.getActions()[0].type).toEqual('GET_ALL_PHOTOS_ERROR');
        expect(store.getActions()[0].errorMessage).toEqual('Cannot fetch all photos');
      });
  });
});
