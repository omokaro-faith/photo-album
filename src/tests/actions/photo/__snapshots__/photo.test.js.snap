// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async getPhotos gets photos 1`] = `
Array [
  Object {
    "photos": Array [],
    "type": "GET_PHOTOS",
  },
]
`;
